### PR TITLE
check if the compiler path and debugger path exist.

### LINKED ISSUE #8836: Misleading error message when GDB is not installed
Bug type: Debugger

If I don't have GDB installed (yet), and try to use the 'build and debug active file' feature (on Linux, with g++), I get an error message: 

image

However, all I have at this point is `tasks.json` (not `launch.json`, which it refs to), and there is no reference to `miDebuggerPath` in it.

It would be useful if this error message provided enough information to determine that the issue is that GDB needs to be installed.

### PR DESCRIPTION
**Pull request by elahehrashedi**
bug fix: https://github.com/microsoft/vscode-cpptools/issues/8836
Do not create a debug configuration for invalid compiler paths.

### CODE REVIEW

File: Extension/src/Debugger/configurationProvider.ts [hunk 1]
```diff
@@ -410,6 +410,15 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
                 const definition: CppBuildTaskDefinition = task.definition as CppBuildTaskDefinition;
                 const compilerPath: string = definition.command;
+                // Filter out the tasks that has an invalid compiler path.
+                const compilerPathExists: boolean = path.isAbsolute(compilerPath) ?
+                    // Absolute path, just check if it exists
+                    await util.checkFileExists(compilerPath) :
+                    // Non-absolute. Check on $PATH
+                    ((await util.whichAsync(compilerPath)) !== undefined);
+                if (!compilerPathExists) {
+                    console.log("Unable to find {0}. {1} task is ignored.", compilerPath, definition.label);
```
Comment (sean-mcmanus) [inline Extension/src/Debugger/configurationProvider.ts]: Is this the correct way to report the error? The string isn't localized. I see other code using stuff like `logger.getOutputChannelLogger().showErrorMessage(localize(`. I think our existing uses of console.log are also incorrect/unideal, but we haven't gotten around to improving it.

**Review by sean-mcmanus — APPROVED** [review]

**Comment by elahehrashedi** [comment]
@Colengc Here I stopped showing the debug configurations in which the debugger is invalid in the drop down. And the user will see the warning in the output channel to know that the debug config was being ignored. Is this a good approach?

**Comment by elahehrashedi** [comment]
@sean-mcmanus does this message look good?
('debugger.path.not.exists', "Unable to find the {0} debugger in this path: {1}. The debug configuration for {2} is ignored.", `\"${debuggerName}\"`, `\"${debuggerPath}\"`, compilerName)

File: Extension/src/Debugger/configurationProvider.ts [hunk 2]
```diff
@@ -467,22 +476,23 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
-                        if (isWindows) {
-                            newConfig.miDebuggerPath = debuggerPath;
-                            return resolve(newConfig);
-                        } else {
-                            fs.stat(debuggerPath, (err, stats: fs.Stats) => {
-                                if (!err && stats && stats.isFile()) {
-                                    newConfig.miDebuggerPath = debuggerPath;
-                                } else {
-                                    newConfig.miDebuggerPath = path.join("/usr", "bin", debuggerName);
+                        if (!await util.checkFileExists(debuggerPath)) {
```
Comment (michelleangela) [inline Extension/src/Debugger/configurationProvider.ts]: Could you clarify why it checks `await util.checkFileExists(debuggerPath)` as false at line 479 and then at line 481 it checks it again as true when there doesn't seem to be a change to `debuggerPath` that is being verified? 

It seems conditional block 479 could be refactored?

File: Extension/src/Debugger/configurationProvider.ts [hunk 3]
```diff
@@ -467,22 +476,23 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+                            if (!isWindows) {
+                                const newDebuggerPath: string = path.join("/usr", "bin", debuggerName);
+                                if (await util.checkFileExists(debuggerPath)) {
```
Comment (sean-mcmanus) [inline Extension/src/Debugger/configurationProvider.ts]: It seems like this should be `if (await util.checkFileExists(newDebuggerPath)) {`

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]

**Comment by sean-mcmanus** [comment]
> @sean-mcmanus does this message look good? ('debugger.path.not.exists', "Unable to find the {0} debugger in this path: {1}. The debug configuration for {2} is ignored.", `\"${debuggerName}\"`, `\"${debuggerPath}\"`, compilerName)

Sure.

File: Extension/src/Debugger/configurationProvider.ts (hunk 2 above)
Comment (elahehrashedi) [reply Extension/src/Debugger/configurationProvider.ts]: line 481 had to use `newDebuggerPath`, it is fixed now.

File: Extension/src/Debugger/configurationProvider.ts [hunk 4]
```diff
@@ -467,22 +480,29 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+                            // Check if debuggerName exists on $PATH
+                            newConfig.miDebuggerPath = debuggerName;
+                            return resolve(newConfig);
                         } else {
-                            fs.stat(debuggerPath, (err, stats: fs.Stats) => {
-                                if (!err && stats && stats.isFile()) {
-                                    newConfig.miDebuggerPath = debuggerPath;
-                                } else {
-                                    newConfig.miDebuggerPath = path.join("/usr", "bin", debuggerName);
+                            if (!isWindows) {
```
Comment (michelleangela) [inline Extension/src/Debugger/configurationProvider.ts]: Nit: `if (!isWindows)` can be part of the `else` on line 492.

File: Extension/src/Debugger/configurationProvider.ts [hunk 5]
```diff
@@ -467,22 +480,29 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+                                const newDebuggerPath: string = path.join("/usr", "bin", debuggerName);
```
Comment (michelleangela) [inline Extension/src/Debugger/configurationProvider.ts]: Nit: recommend to rename `newDebuggerPath` to `usrDebuggerPath` to more represent the actual path.

**Review by michelleangela — APPROVED** [review]

File: Extension/src/Debugger/configurationProvider.ts (hunk 4 above)
Comment (elahehrashedi) [reply Extension/src/Debugger/configurationProvider.ts]:
```C++
                        } else {
                            if (!isWindows) {
                                const usrDebuggerPath: string = path.join("/usr", "bin", debuggerName);
                                if (await util.checkFileExists(usrDebuggerPath)) {
                                    newConfig.miDebuggerPath = usrDebuggerPath;
                                    return resolve(newConfig);
                                }
                            }
                            logger.getOutputChannelLogger().appendLine(localize('debugger.path.not.exists', "Unable to find the {0} debugger. The debug configuration for {2} is ignored.", `\"${debuggerName}\"`, compilerName));
                            return resolve(undefined);
                        }
```

The else has two parts, `iswindows `and `!inswindows`.
I simplify it .

**Review by michelleangela — APPROVED** [review]

**Review by michelleangela — APPROVED** [review]

**Review by michelleangela — APPROVED** [review]

File: Extension/src/Debugger/configurationProvider.ts [hunk 6]
```diff
@@ -408,9 +408,18 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
         // Generating a task is async, therefore we must *await* *all* map(task => config) Promises to resolve.
         let configs: CppDebugConfiguration[] = [];
         if (buildTasks.length !== 0) {
-            configs = await Promise.all(buildTasks.map<Promise<CppDebugConfiguration>>(async task => {
+            configs = (await Promise.all(buildTasks.map<Promise<CppDebugConfiguration|undefined>>(async task => {
```
Comment (sean-mcmanus) [inline Extension/src/Debugger/configurationProvider.ts]: I think we normally surround the | with spaces.

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]

**Review by sean-mcmanus — APPROVED** [review]

File: Extension/src/Debugger/configurationProvider.ts [hunk 7]
```diff
@@ -439,20 +448,24 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
                 const isCl: boolean = compilerName === "cl.exe";
                 newConfig.cwd = isWindows && !isCl && !process.env.PATH?.includes(path.dirname(compilerPath)) ? path.dirname(compilerPath) : "${fileDirname}";
 
-                return new Promise<CppDebugConfiguration>(resolve => {
+                return new Promise<CppDebugConfiguration|undefined>(async resolve => {
```
Comment (sean-mcmanus) [inline Extension/src/Debugger/configurationProvider.ts]: Can you add spaces around | here too?

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]

**Review by sean-mcmanus — APPROVED** [review]

**Review by sean-mcmanus — APPROVED** [review]

**Pull request merged**